### PR TITLE
Support load sensitive PoolWeftExecutorService

### PR DESCRIPTION
**Pull request by ruhan1**

### CODE REVIEW

File: src/main/java/org/commonjava/cdi/util/weft/LoadSensitivePoolWeftExecutorService.java [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+        return delegate.submit( callable );
+    }
+
+    @Override
+    public <T> Future<T> submit( Runnable runnable, T t )
+    {
+        if ( !isHealthy() )
+        {
+            throw new PoolOverloadException();
```
Comment (jdcasey) [inline src/main/java/org/commonjava/cdi/util/weft/LoadSensitivePoolWeftExecutorService.java]: We should use the pool name here, and maybe give stats like the load factor and pool size here in the constructor. That would help with advice the caller might pass back to its upstream users.

Comment (ruhan1) [reply src/main/java/org/commonjava/cdi/util/weft/LoadSensitivePoolWeftExecutorService.java]: OK

**Comment by ruhan1** [comment]
Updated. Put this functionality in the PoolWeftExecutorService.

File: src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java [hunk 2]
```diff
@@ -131,79 +135,144 @@ public boolean awaitTermination( long l, TimeUnit timeUnit )
...
+        return loadSensitive && !isHealthy();
+    }
+    
     @Override
     public <T> Future<T> submit( Callable<T> callable )
     {
+        if ( isOverloaded() )
+        {
+            throw new PoolOverloadException( getName(), getLoadFactor(), getCurrentLoad(), getThreadCount() );
+        }
```
Comment (sswguo) [inline src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java]: How about move the throw clause into the function too ?

Comment (ruhan1) [reply src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java]: Sounds good.

File: src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java [hunk 3]
```diff
@@ -131,79 +135,111 @@ public boolean awaitTermination( long l, TimeUnit timeUnit )
         return delegate.awaitTermination( l, timeUnit );
     }
 
+    private void verifyLoad()
+    {
+        if ( loadSensitive && !isHealthy() )
+        {
+            throw new PoolOverloadException( getName(), getLoadFactor(), getCurrentLoad(), getThreadCount() );
```
Comment (ligangty) [inline src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java]: One question here: how will we control this exception in indy? Should we catch this exception in indy's calling points of all WeftExcecutorService and handle them separately?

Comment (jdcasey) [reply src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java]: What about making this exception extend Executioner option? Is that in the method signature?

Comment (jdcasey) [reply src/main/java/org/commonjava/cdi/util/weft/PoolWeftExecutorService.java]: I'm resolving this, so I can merge it and try to modify things slightly so it's a little easier to integrate into Indy. We may end up changing this later, but I want to get something built and start getting some feedback / mileage on it.

**Comment by ruhan1** [comment]
The submit/execute throws no exception. Apparently, Java API does not consider the load problem. If we make this a checked exception, it will break the rule. Java can not throw additional checked exceptions more than the interface (but it may throw fewer). I think we don't have much choice. We either catch/handle the runtime one, or let it thrown to the upper level.

**Pull request merged**